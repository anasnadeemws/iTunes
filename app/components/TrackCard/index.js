/**
 *
 * TrackCard
 *
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Card } from '@mui/material';
import styled from '@emotion/styled';
import T from '@components/T';
import If from '@components/If';
import isEmpty from 'lodash/isEmpty';
import Box from '@mui/material/Box';
import CardMedia from '@mui/material/CardMedia';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import PlayArrowIcon from '@mui/icons-material/PlayArrow';
import PauseIcon from '@mui/icons-material/Pause';

const TrackCustomCard = styled(Card)`
  && {
    display: flex;
    margin: 1rem;
    width: 32rem;
    justify-content: space-between;
    :hover {
      box-shadow: 5;
    }
    transition: all 0.25s linear;
  }
`;

const TrackContentBox = styled(Box)`
  && {
    display: flex;
    flex-direction: column;
    padding: 1rem;
  }
`;

const TrackMedia = styled(CardMedia)`
  && {
    width: 10rem;
  }
`;

const TrackContentHeaderBox = styled(TrackContentBox)`
  && {
    flex-direction: row;
    align-items: center;
    padding-bottom: 1rem;
  }
`;

const TrackPauseIcon = styled(PauseIcon)`
  && {
    height: 2rem;
    width: 2rem;
  }
`;

const TrackPlayIcon = styled(PlayArrowIcon)`
  && {
    height: 2rem;
    width: 2rem;
  }
`;

export function TrackCard({ collectionName, artistName, shortDescription, artworkUrl100, previewUrl }) {
  const [playTrack, setPlayTrack] = useState(false);

  // Helpers
  const truncateWord = (word, truncateLen) => {
    if (!word) {
      return '';
    }
    return word.length > truncateLen ? `${word.substring(0, truncateLen)}...` : word;
  };

  return (
    <TrackCustomCard data-testid="track-card">
      <TrackContentBox>
        <TrackContentHeaderBox>
          <If condition={!isEmpty(previewUrl)}>
            <IconButton aria-label="play/pause" onClick={() => setPlayTrack(!playTrack)}>
              <If condition={playTrack} otherwise={<TrackPlayIcon />}>
                <TrackPauseIcon />
              </If>
            </IconButton>
          </If>
          <Typography component="div" variant="h5">
            <If
              condition={!isEmpty(collectionName)}
              otherwise={<T data-testid="collection_name_unavailable" id="collection_name_unavailable" />}
            >
              <T data-testid="collectionName" id="collection_name" values={{ collectionName: collectionName }} />
            </If>
          </Typography>
        </TrackContentHeaderBox>
        <Typography variant="subtitle1" color="text.secondary" component="div">
          <If
            condition={!isEmpty(artistName)}
            otherwise={<T data-testid="track_artist_name_unavailable" id="track_artist_name_unavailable" />}
          >
            <T data-testid="artistName" id="track_artist_name" values={{ artistName: artistName }} />
          </If>
        </Typography>
        <If
          condition={!isEmpty(shortDescription)}
          otherwise={<T data-testid="track_shortdesc_unavailable" id="track_shortdesc_unavailable" />}
        >
          <T data-testid="track_shortdesc" id="track_shortdesc" values={{ desc: truncateWord(shortDescription, 38) }} />
        </If>
      </TrackContentBox>
      <If condition={!isEmpty(artworkUrl100)}>
        <TrackMedia component="img" image={artworkUrl100} alt="Poster unavailable" />
      </If>
      <If condition={playTrack}>
        <audio autoPlay name="media">
          <source src={previewUrl} />
        </audio>
      </If>
    </TrackCustomCard>
  );
}

TrackCard.propTypes = {
  artistName: PropTypes.string,
  artworkUrl100: PropTypes.string,
  collectionName: PropTypes.string,
  shortDescription: PropTypes.string,
  previewUrl: PropTypes.string
};

export default TrackCard;
