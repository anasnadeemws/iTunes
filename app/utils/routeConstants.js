export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  tracks: {
    route: '/tracks',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  track: {
    route: '/track:trackid',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
};
