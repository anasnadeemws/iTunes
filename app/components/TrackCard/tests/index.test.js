/**
 *
 * Tests for TrackCard
 *s
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import TrackCard from '../index';
import { translate } from '@app/utils/index';

describe('<TrackCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackCard />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackCard component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackCard />);
    expect(getAllByTestId('track-card').length).toBe(1);
  });

  it('should render the track details inside the card', () => {
    const collectionName = 'Sunflower';
    const artistName = 'Post Malone';
    const shortDescription = '';
    const artworkUrl100 = 'https://someurl.com';
    const previewUrl = '';
    const { getByTestId } = renderWithIntl(
      <TrackCard
        collectionName={collectionName}
        artistName={artistName}
        shortDescription={shortDescription}
        artworkUrl100={artworkUrl100}
        previewUrl={previewUrl}
      />
    );
    expect(getByTestId('collectionName')).toHaveTextContent(collectionName);
  });

  it('should render the repository unavailable messages in case any props are unavailable or have falsy values', () => {
    const repoUnavailable = translate('repo_name_unavailable');
    const fullNameUnavailable = translate('repo_full_name_unavailable');
    const stargazersUnavailable = translate('repo_stars_unavailable');
    const { getByTestId } = renderWithIntl(<TrackCard />);
    expect(getByTestId('name-unavailable')).toHaveTextContent(repoUnavailable);
    expect(getByTestId('fullName-unavailable')).toHaveTextContent(fullNameUnavailable);
    expect(getByTestId('stargazers-unavaiable')).toHaveTextContent(stargazersUnavailable);
  });
});
