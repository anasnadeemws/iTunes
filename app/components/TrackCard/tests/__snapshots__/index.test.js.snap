// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e12nrugs5 css-b8f0w9-MuiPaper-root-MuiCard-root-TrackCustomCard"
    data-testid="track-card">
    <div class="e12nrugs4 MuiBox-root css-14ywcf9-TrackContentBox">
        <div
          class="e12nrugs2 MuiBox-root css-3wx8pt-TrackContentBox-TrackContentHeaderBox"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
            aria-label="play/pause"
          >
            <svg
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium e12nrugs0 css-habosj-MuiSvgIcon-root-TrackPlayIcon"
              focusable="false"
              aria-hidden="true"
              viewBox="0 0 24 24"
              data-testid="PlayArrowIcon"
            >
              <path d="M8 5v14l11-7z"></path></svg
            ><span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            ></span>
          </button>
          <div
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            <p
              data-testid="collection_name_unavailable"
              class="css-1c83g7r-StyledText-regular-normal-standard egtqi0h0"
            >
              Track name is unavailable
            </p>
          </div>
        </div>
        <div
          class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
        >
          <p
            data-testid="track_artist_name_unavailable"
            class="css-1c83g7r-StyledText-regular-normal-standard egtqi0h0"
          >
            Artist name is unavailable
          </p>
        </div>
        <p
          data-testid="track_shortdesc_unavailable"
          class="css-1c83g7r-StyledText-regular-normal-standard egtqi0h0"
        >
          Description unavaiable
        </p>
      </div>
    </div>
  </div>
</body>
`;
